gem/optimise.py

```diff
@@ -230,13 +230,47 @@ def delta_elimination(sum_indices, factors):
     return sum_indices, [e for e in factors if e != one]
 
 
+def associate_product(factors):
```

Comment:
Probably, but there is no obvious place to do that, so I would just make a separate, DAG-safe pass for that.